src/language/branches.md

```diff
@@ -0,0 +1,295 @@
+  }
+  ```  
+- **Variable patterns**: the matched value binds to a new variable.
+  ```dart
+  switch ((1, 2)) {
+    case (var a, var b): ...     // Matches if 1 and 2 can be assigned to a and b
+  }
+  ```  
+- **Logical** or **relational patterns**: the object matches if the comparison to a constant returns `true`.
+  ```dart
+  switch (code) {
+    case >= first && <= last: ...    
+    // Matches if code's value is between the constant values of first and last
+  }
+  ```  
+- **Null check** or **assert patterns**: the match is first on whether the object is not null, then on the value.
+  ```dart
+  switch (maybeString) {
+    case var s?:                 
+    // Matches if maybe is not null, and then against the inner pattern "var s"
+  }
+  ```  
+- **Object patterns**: the match exposes getters on the object type.
+  ```dart
+  switch (Shape shape) {
+    case Rect(width: var w, height: var h): ...
+    // Matches if shape is of type Rect, and then against the properties of Rect
+  }
+  ```  
+
+Values destructured by a case pattern become local variables, only available in the
+scope of that clause. 
+
+### Switch statements
+
+Non-empty `case` clauses implicitly jump to the end of the switch after completion.
+Other valid ways to end a non-empty `case` clause are a [`continue`][break],
+[`throw`][], or [`return`][] statement.
+
+Use a `default` or wildcard `_` clause to execute code when no `case` clause matches:
+
+```dart
+var command = 'OPEN';
+switch (command) {
+  case 'CLOSED':
+    executeClosed();
+  case 'PENDING':
+    executePending();
+  case 'APPROVED':
+    executeApproved();
+  case 'DENIED':
+    executeDenied();
+  case 'OPEN':
+    executeOpen();
+  default:
+    executeUnknown();
+}
+```
+
+Empty cases fall through to the next case. For an empty case that does not fall
+through, use [`break`][break] for its body:
+
+```dart
+var command = 'CLOSED';
+switch (command) {
+  case 'CLOSED': // Empty case falls through.
+  case 'NOW_CLOSED':
+    // Runs for both CLOSED and NOW_CLOSED.
+    executeNowClosed();
+}
+```
+
+For non-sequential fall-through, you can use a [`continue` statement][break] and
+a label:
+
+```dart
+var command = 'CLOSED';
+switch (command) {
+  case 'CLOSED':
+    executeClosed();
+    continue nowClosed;
+  // Continues executing at the nowClosed label.
+
+  nowClosed:
+  case 'NOW_CLOSED':
+    // Runs for both CLOSED and NOW_CLOSED.
+    executeNowClosed();
+}
+```
+
+### Switch expressions
+
+_Switch expressions_ write a result value determined by a switch
+directly into a variable or return statement. 
+
+The syntax rules for a switch expression are:
+- The expression body is a series of case patterns.
+- Cases _do not_ use the `case` keyword.
+- Case patterns are followed by `=>`, and then a case body. 
+- Case bodies terminate with a comma `,`.
+- The switch expression body terminates with a semicolon `;`.
+- The wildcard pattern `_` catches remaining values instead of `default`.
```

Comment:
You can actually use `_` as a default case in switch statements too. (Whether that will become idiomatic or not is anybody's guess.)

Maybe something like:

> Switch expression syntax is slightly different from switch statements:
>
> - Cases *don't* start with the `case` keyword.
> - The case pattern is separated from its body using `=>` instead of `:`.
> - The case body is a single expression instead of a series of statements. Every case must have a body -- there is no implicit fallthrough for empty cases.
> - Cases are separated by `,` (and an optional trailing `,` is allowed).